wdl-grammar/src/experimental/diagnostic.rs

```diff
@@ -0,0 +1,260 @@
+//! Definition of diagnostics displayed to users.
+
+use std::fmt;
+
+use rowan::TextRange;
+
+/// Represents a span of source.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Span {
+    /// The start of the span.
+    start: usize,
+    /// The end of the span.
+    end: usize,
+}
+
+impl Span {
+    /// Creates a new span from the given start and end.
+    pub fn new(start: usize, len: usize) -> Self {
+        Self {
+            start,
+            end: start + len,
+        }
+    }
+
+    /// Gets the start of the span.
+    pub fn start(&self) -> usize {
+        self.start
+    }
+
+    /// Gets the end of the span.
+    pub fn end(&self) -> usize {
+        self.end
+    }
+
+    /// Gets the length of the span.
+    pub fn len(&self) -> usize {
+        self.end - self.start
+    }
+
+    /// Determines if the span is empty.
+    pub fn is_empty(&self) -> bool {
+        self.start == self.end
+    }
+}
+
+impl fmt::Display for Span {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{start}..{end}", start = self.start, end = self.end)
+    }
+}
+
+impl From<logos::Span> for Span {
+    fn from(value: logos::Span) -> Self {
+        Self::new(value.start, value.len())
+    }
+}
+
+/// Represents the severity of a diagnostic.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum Severity {
```

Comment:
I think the notion of `Severity` still needs some workshopping. Continuing the conversation from #12, here's my idea for `Severity`:
- `Error` will be anything and everything coming out of `wdl-grammar` and `wdl-ast`
    - it means "something is _wrong_ with the provided code" and "it violates the WDL specification"
    - `wdl-lint` should never emit an `Error`
    - Is there a time where `wdl-grammar` or `wdl-ast` should emit a `Warning` or `Note`? I think not.
- `Warning` this is emitted by `wdl-lint`
    - It means "this is not 'good' WDL. A best practice has been ignored."
    - I've changed my mind from a previous position and now think every lint violation should be a `Warning`
    - In `sprocket lint` these should have a non-zero exit code
- `Note` - I don't see a purpose for `Note`
    - Maybe I just need a good example pointed out to me though 🤷

Reply:
Keeping in mind that `note`, `warning`, and `error` all colorize differently in normal terminal output, I still see value in `note` conveying what we used to convey with a `Low` lint level: this particular lint isn't all that important, but it is valuable for style and consistency.

It would convey to users that the lint is more opinion than "something's potentially wrong here, please investigate" and would be a reasonable thing to disable if one doesn't agree with the lint.

> Is there a time where `wdl-grammar` or `wdl-ast` should emit a Warning or Note? I think not.

Real-world compilers emit warnings and I don't want to limit this for the future or make the tooling "linting-oriented"; while the parser itself might not emit warnings, certainly a future WDL evaluator would emit warnings (e.g. "unused variable", "unreachable code", etc.) in addition to errors (e.g. "unknown name reference", "type mismatch", etc.) and I don't want to have to use a different `Diagnostic` type for that either.

We should also keep in mind that linting is _opt-in_ too; there will be plenty of other tools we may want to build in the future that will produce warnings and errors (and maybe even notes).

Reply:
Good points! Making me put my thinking cap on 🤠 (I wear a cowboy hat when deep in thought)

> Real-world compilers emit warnings and I don't want to limit this for the future or make the tooling "linting-oriented"; while the parser itself might not emit warnings, certainly a future WDL evaluator would emit warnings (e.g. "unused variable", "unreachable code", etc.) in addition to errors (e.g. "unknown name reference", "type mismatch", etc.) and I don't want to have to use a different Diagnostic type for that either.
> 
> We should also keep in mind that linting is opt-in too; there will be plenty of other tools we may want to build in the future that will produce warnings and errors (and maybe even notes).

This has convinced me to lighten up on my position. I initially wanted to see this `Severity` `enum` to only have `Error` and `Warning`, but you're right that there are plenty of use cases for `Note`. _But_, I don't think any of those `Note` uses are being implemented today/in this PR. I still feel all lints, even the whitespace ones, belong at the `Warning` level.

> this particular lint isn't all that important, but it is valuable for style and consistency.
> It would convey to users that the lint is more opinion than "something's potentially wrong here, please investigate" and would be a reasonable thing to disable if one doesn't agree with the lint.

I was clearly at one point in favor of having a `Low` category, and by extension a past version of me would be agreeing with this logic. But @adthrasher changed my mind.

We should keep our ultimate goal in mind. We _are_ opinionated. We think these lints are the best way to write WDL. We want people to adopt the lints we're implementing. The current state of community WDL is _a problem we're trying to **fix**_.

If we want people to adopt our lints, why would we say some of them are less important? It feels like encouraging them to disable certain lints. I'd rather present a unified front, that all of these lints matter and should be adopted.

Reply:
> If we want people to adopt our lints, why would we say some of them are less important? It feels like encouraging them to disable certain lints. I'd rather present a unified front, that all of these lints matter and should be adopted.

I still have some (mild) concerns over treating all lint rules as having the same importance, and also a default of "non-zero exit on warning", as being _barriers_ to adoption rather than _promoters_ of adoption, though.

I've turned on new linting tools on large, existing source bases before and it can be _a lot_ of work to get a linter happy. Sometimes knowing that a thing the linter is complaining about isn't all that important (vs ones that need to be immediately fixed) and disabling it _for now_ is the right call to make to progress towards the goal of being linter "clean".

With that all said, I propose we keep `Diagnostic` with the severity levels for future flexibility, but change all of the lint rules to `Diagnostic::warning` to meet our goal of promoting consistency to the larger WDL community.

We can always revisit that for the future based on user feedback.

Reply:
That's a valid perspective! I think we're both bringing real concerns to the table.

> We can always revisit that for the future based on user feedback.

Yes, I meant to include something to that effect. We don't need to feel tied to any decision we make today. I would go so far as to say we need to keep ourselves open to the community's feedback on this.

> I've turned on new linting tools on large, existing source bases before and it can be a lot of work to get a linter happy. Sometimes knowing that a thing the linter is complaining about isn't all that important (vs ones that need to be immediately fixed) and disabling it for now is the right call to make to progress towards the goal of being linter "clean".

I've been there, and that is one reason I consider it a high priority to get `sprocket fmt` implemented.

With all this in mind, I'd be happy charging ahead with all lint's being at the `Warning` level. It will be trivial to update pending more input on the issue.

Reply:
I've updated them to the warning level.

Reply:
>   * Is there a time where `wdl-grammar` or `wdl-ast` should emit a `Warning` or `Note`? I think not.

Peter highlighted a couple of places below. Another is language features slated for deprecation. They are valid for the current version, but the parser should `warn` or `note` those to the user.